src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -56,75 +58,83 @@ internal class RenamerProjectTreeActionHandler : ProjectTreeActionHandlerBase, I
             _vsOnlineServices = vsOnlineServices;
             _threadingService = threadingService;
             _extensibility = extensibility;
+            _languageServiceDone = new Dictionary<(string, string), SemaphoreSlim>();
         }
 
         public void HandleRename(string oldFilePath, string newFilePath)
         {
-            if (_oldFilePath == oldFilePath)
+            string newFileWithExtension = Path.GetFileName(newFilePath);
+            if (_languageServiceDone.TryGetValue((oldFilePath, newFileWithExtension), out SemaphoreSlim semLanguegeServiceDone))
             {
-                _isLanguageServiceDone = true;
+                semLanguegeServiceDone.Release();
             }
         }
 
         public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
         {
-            Assumes.Present(context);
-            Assumes.Present(node);
-            Assumes.Present(value);
+            Requires.NotNull(context, nameof(Context));
+            Requires.NotNull(node, nameof(node));
+            Requires.NotNullOrEmpty(value, nameof(value));
 
-            _oldFilePath = node.FilePath;
-            string? newNameWithExtension = value;
             // These variables are need to synchronize with Roslyn
-            _isLanguageServiceDone = false;
+            string oldFilePath = node.FilePath;
+            string newFileNameWithExtension = value;
 
             // Do not offer to rename the file in VS Online scenarios.
             if (_vsOnlineServices.ConnectedToVSOnline)
             {
                 return;
             }
 
-            if (await IsAutomationFunction() || node.IsFolder)
+            if (await IsAutomationFunctionAsync() || node.IsFolder)
             {
                 // Do not display rename Prompt
                 await base.RenameAsync(context, node, value);
                 return;
             }
 
             // Get the list of possible actions to execute
-            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
             CodeAnalysis.Project? project = GetCurrentProject();
             if (project is null)
             {
                 return;
             }
 
-            CodeAnalysis.Document? oldDocument = GetDocument(project, _oldFilePath);
+            CodeAnalysis.Document? oldDocument = GetDocument(project, oldFilePath);
             if (oldDocument is null)
             {
                 return;
             }
 
-            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newNameWithExtension, oldDocument.Folders);
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newFileNameWithExtension);
```

Comment:
If you don't pass the new folders here - how does the renamer know to fix up the namespace?

Reply:
It was using `oldDocument.Folders` before, which would always be the same as passing nothing. If the new namespace should be fixed here too (my assumption was that was coming later) then it needs `newDocument.Folders`